merkle/log_verifier_test.go

```diff
@@ -0,0 +1,448 @@
+	for i := 1; i < 6; i++ {
+		// Construct the path.
+		proof := [][]byte{}
+		for j := int64(0); j < inclusionProofs[i].proofLength; j++ {
+			proof = append(proof, inclusionProofs[i].proof[j].h)
+		}
+		err := verifierCheck(&v, inclusionProofs[i].leaf-1, inclusionProofs[i].snapshot, proof,
+			roots[inclusionProofs[i].snapshot-1].h,
+			inputs[inclusionProofs[i].leaf-1].h)
+		if err != nil {
+			t.Fatalf("i=%d: %s", i, err)
+		}
+	}
+
+}
+
+func TestVerifyConsistencyProof(t *testing.T) {
+	v := getVerifier()
+
+	proof := [][]byte{}
+	root1 := []byte("don't care")
+	root2 := []byte("don't care")
+
+	// Snapshots that are always consistent
+	if err := verifierConsistencyCheck(&v, 0, 0, root1, root2, proof); err != nil {
+		t.Fatalf("Failed to verify proof: %s", err)
+	}
+	if err := verifierConsistencyCheck(&v, 0, 1, root1, root2, proof); err != nil {
+		t.Fatalf("Failed to verify proof: %s", err)
+	}
+	if err := verifierConsistencyCheck(&v, 1, 1, root1, root2, proof); err != nil {
+		t.Fatalf("Failed to verify proof: %s", err)
+	}
+
+	// Invalid consistency proofs.
+	// Time travel to the past.
+	if err := verifierConsistencyCheck(&v, 1, 0, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified timetravelling proof")
+	}
+	if err := verifierConsistencyCheck(&v, 2, 1, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified timetravelling proof")
+	}
+
+	// Empty proof
+	if err := verifierConsistencyCheck(&v, 1, 2, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified timetravelling proof")
+	}
+
+	root1 = dh(sha256EmptyTreeHash)
+	// Roots don't match.
+	if err := verifierConsistencyCheck(&v, 0, 0, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified mismatched root")
+	}
+	if err := verifierConsistencyCheck(&v, 1, 1, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified mismatched root")
+	}
+
+	// Roots match but the proof is not empty.
+	root2 = dh(sha256EmptyTreeHash)
+	proof = [][]byte{dh(sha256EmptyTreeHash)}
+
+	if err := verifierConsistencyCheck(&v, 0, 0, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified non-empty proof")
+	}
+	if err := verifierConsistencyCheck(&v, 0, 1, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified non-empty proof")
+	}
+	if err := verifierConsistencyCheck(&v, 1, 1, root1, root2, proof); err == nil {
+		t.Fatal("Incorrectly verified non-empty proof")
+	}
+
+	// Known good proofs.
+	roots := getRoots()
+	proofs := getConsistencyProofs()
+	for i := 0; i < 4; i++ {
+		proof := [][]byte{}
+		for j := int64(0); j < proofs[i].proofLen; j++ {
+			proof = append(proof, proofs[i].proof[j].h)
+		}
+		snapshot1 := proofs[i].snapshot1
+		snapshot2 := proofs[i].snapshot2
+		err := verifierConsistencyCheck(&v, snapshot1, snapshot2,
+			roots[snapshot1-1].h,
+			roots[snapshot2-1].h, proof)
+		if err != nil {
+			t.Fatalf("Failed to verify known good proof for i=%d: %s", i, err)
+		}
+	}
+}
+
+func dh(h string) []byte {
```

Comment:
This exists in testonly package as MustHexDecode. Can use that unless this code needs to be standalone, which it might.